Zotero/Scenes/Detail/Trash/ViewModels/TrashActionHandler.swift

```diff
@@ -0,0 +1,663 @@
+        if let collections, collectionsIdx < collections.count {
+            while collectionsIdx < collections.count {
+                keys.append(TrashKey(type: .collection, key: collections[collectionsIdx].key))
+                keyToIdx[keys.last!] = collectionsIdx
+                collectionsIdx += 1
+            }
+        }
+        if let items, itemsIdx < items.count {
+            while itemsIdx < items.count {
+                keys.append(TrashKey(type: .item, key: items[itemsIdx].key))
+                keyToIdx[keys.last!] = itemsIdx
+                itemsIdx += 1
+            }
+        }
+        return (keys, keyToIdx)
+    }
+
+    private func areInIncreasingOrder(lObject: TrashObject, rObject: TrashObject, sortType: ItemsSortType) -> Bool {
+        let initialResult: ComparisonResult
+
+        switch sortType.field {
+        case .creator:
+            initialResult = compare(lValue: lObject.creatorSummary, rValue: rObject.creatorSummary)
+
+        case .date:
+            initialResult = compare(lValue: lObject.date, rValue: rObject.date)
+
+        case .dateAdded:
+            initialResult = compare(lValue: lObject.dateAdded, rValue: rObject.dateAdded)
+
+        case .dateModified:
+            initialResult = compare(lValue: lObject.dateModified, rValue: rObject.dateModified)
+
+        case .itemType:
+            initialResult = compare(lValue: lObject.sortType, rValue: rObject.sortType)
+
+        case .publicationTitle:
+            initialResult = compare(lValue: lObject.publicationTitle, rValue: rObject.publicationTitle)
+
+        case .publisher:
+            DDLogInfo("LPublisher: \(lObject.key); '\(lObject.publisher ?? "nil")' - \(rObject.key); '\(rObject.publisher ?? "nil")'")
+            initialResult = compare(lValue: lObject.publisher, rValue: rObject.publisher)
+
+        case .year:
+            initialResult = compare(lValue: lObject.year, rValue: rObject.year)
+
+        case .title:
+            return isInIncreasingOrder(result: compare(lValue: lObject.sortTitle, rValue: rObject.sortTitle), ascending: sortType.ascending, comparedSame: nil)
+        }
+
+        return isInIncreasingOrder(result: initialResult, ascending: sortType.ascending, comparedSame: { compare(lValue: lObject.sortTitle, rValue: rObject.sortTitle) })
+
+        func isInIncreasingOrder(result: ComparisonResult, ascending: Bool, comparedSame: (() -> ComparisonResult)?) -> Bool {
+            switch result {
+            case .orderedSame:
+                if let result = comparedSame?() {
+                    return ascending ? result == .orderedAscending : result == .orderedDescending
+                }
+                return true
+
+            case .orderedAscending:
+                return ascending
+
+            case .orderedDescending:
+                return !ascending
+            }
+        }
+
+        func compare<Val>(lValue: Val?, rValue: Val?, nonNilCompare: (Val, Val) -> ComparisonResult) -> ComparisonResult {
```

Comment:
It may be more clear to use the already available constants for `nil` alternatives. E.g. for `Int`, since we want to put `nil` values in the end of an ascending order, we may use instead:
```
func compare(lValue: Int?, rValue: Int?) -> ComparisonResult {
    let safeLValue = lValue ?? .max
    let safeRValue = rValue ?? .max
    if safeLValue == safeRValue {
        return .orderedSame
    }
    return safeLValue < safeRValue ? .orderedAscending : .orderedDescending
}
```

Similarly for `Date` we can use `.distantFuture`.

If these comparisons are to be reused throughout the app, we may even consider extending their types accodingly.

Reply:
It's not really usable for `String`, so I'd have to keep original implementation for strings anyway.